### DISCUSSION TITLE
Flow save& Restore

### DISCUSSION THREAD

**Discussion by ftayyab** [xyflow/xyflow#1920] [Q&A]
Hi, 

I have custom nodes and I am able to save and load the flows (made up of custom nodes) in a database and retrieve and load them up. All elements load up correctly but the internal custom controls do not maintain their values. Is it possible for those controls to maintain their values as well automatically. If its not automatic, will it have to be managed manually.

**Comment by moklick** [comment]
Hey! What do you mean with "internal custom controls"?

**Comment by ftayyab** [comment]
Apologies, if it was not clear. 

If I have a custom Node Type, in the case below a select/drop down control on a custom Node. I am able to save the state of the Node using toObjects(). However on restoring the elements, the selection on the select control is not maintained and is lost. I think I have to manage it myself, but I wanted to confirm, if that is expected behaviour or is there something I am missing.

 

```js 
<div style={customNodeStyles}>
      <div style={headingStyle}>
        <div style={headingIcon}>
          <ListItemIcon sx={{ color: 'inherit' }}>
            <LensBlurIcon/>
          </ListItemIcon>
        </div>
        <div style={headingText}>
          Input Layer
        </div>
      </div>     

      <Handle
        type="source"
        position="bottom"
      />

      <Box sx={{ minWidth: 100, padding:'0px'}}>
        <FormControl variant="standard" sx={{ m: 1, minWidth: 120, padding:'0px'}}>
          <Select
            label="Table"
            className="nodrag"
            defaultValue = ""
            sx={{fontSize:'0.7rem', color:'#96D9FF', fontFamily: 'Poppins, sans-serif'}}
            onChange={data.handler}
          >
            {layerList}
          </Select>
        </FormControl>
      </Box>
    </div>
```

**Reply by moklick** [reply]
You need to save the select value as well and pass it via the `value` prop (don't use `defaultValue`).

**Reply by ftayyab** [reply]
Hi,

I have been able to save the select value in the data property of the newNode object

`let newNode = {
        id: getId(),
        type,
        position,
        data: { },
      };

newNode.data = {
        ...newNode.data,
        handler: (e)=>{
            let selectedObj = {id: newNode.id, layer: e.target.value};
            newNode.data.selectedLayer = selectedObj;
        },
    };
`

selectedLayer is a state in the parent component and the child component is the newNode. I needed to use state in the child component for it to render the value in the select box and set that in the handler above (not sure if that is the correct way).

On Restoring the elements from the database, I can load up the elements but I am still unable to reference the select control on the loaded element. I have tried using forwardRef but get a null. How can i (on load of the Node) be able to set teh selectedLayer which is now available on data property.

``

**Comment by ftayyab** [comment]
Got it to work. However now having out of range value error for the select. This is because the select items are not loaded when the value is assigned.
